ioctl/cmd/ins/insregister.go

```diff
@@ -0,0 +1,352 @@
+	)
+	_insRegisterCmd.Flags().StringVarP(
+		&resolver, "resolver", "r",
+		"0x41B9132D4661E016A09a61B314a1DFc0038CE3e8",
+		config.TranslateInLang(_registerResolverUsages, config.UILanguage),
+	)
+	_insRegisterCmd.Flags().StringVarP(
+		&owner, "owner", "w",
+		"",
+		config.TranslateInLang(_registerOwnerUsages, config.UILanguage),
+	)
+
+	// init secret
+	sha := sha3.NewLegacyKeccak256()
+	sha.Write([]byte("secret"))
+	hash := sha.Sum(nil)
+	copy(secret[:], hash)
+}
+
+func makeCommitment(
+	name string,
+	owner common.Address,
+	duration *big.Int,
+	secret [32]byte,
+	data [][]byte,
+	reverseRecord bool,
+	ownerControlledFuses uint16,
+) ([32]byte, error) {
+	resolver := common.HexToAddress(resolver)
+	sha := sha3.NewLegacyKeccak256()
+	sha.Write([]byte(name))
+	lable := [32]byte{}
+	copy(lable[:], sha.Sum(nil))
+
+	Bytes32, _ := abi.NewType("bytes32", "", nil)
+	Address, _ := abi.NewType("address", "", nil)
+	Uint256, _ := abi.NewType("uint256", "", nil)
+	BytesArray, _ := abi.NewType("bytes[]", "", nil)
+	Bool, _ := abi.NewType("bool", "", nil)
+	Uint16, _ := abi.NewType("uint16", "", nil)
+	args := abi.Arguments{
+		{Type: Bytes32, Name: "name"},
+		{Type: Address, Name: "owner"},
+		{Type: Uint256, Name: "duration"},
+		{Type: Bytes32, Name: "secret"},
+		{Type: Address, Name: "resolver"},
+		{Type: BytesArray, Name: "data"},
+		{Type: Bool, Name: "reverseRecord"},
+		{Type: Uint16, Name: "ownerControlledFuses"},
+	}
+
+	packed, err := args.Pack(lable, owner, duration, secret, resolver, data, reverseRecord, ownerControlledFuses)
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	hash := sha3.NewLegacyKeccak256()
+	hash.Write(packed)
+	result := [32]byte{}
+	copy(result[:], hash.Sum(nil))
+
+	return result, nil
+}
+
+func commit(name string, data []byte, ioController address.Address, controllerAbi abi.ABI) error {
+	commitment, err := makeCommitment(
+		name,
+		common.HexToAddress(owner),
+		duration,
+		secret,
+		[][]byte{data},
+		true,
+		0,
+	)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("commit commitment for %s: %s ...\n", name, hex.EncodeToString(commitment[:]))
+
+	commitData, err := controllerAbi.Pack("commit", commitment)
+	if err != nil {
+		return output.NewError(output.ConvertError, "failed to pack commit arguments", err)
+	}
+	err = action.Execute(ioController.String(), big.NewInt(0), commitData)
+	if err != nil {
+		return output.NewError(output.UpdateError, "failed to commit commitment", err)
+	}
+	return nil
+}
+
+func register(args []string) error {
+	name := args[0]
+
+	if owner == "" {
+		signer, err := action.Signer()
+		if err != nil {
+			return output.NewError(output.AddressError, "failed to get signer address", err)
+		}
+		signerAddr, err := address.FromString(signer)
+		if err != nil {
+			return output.NewError(output.AddressError, "failed to convert signer address", err)
+		}
+		owner = signerAddr.Hex()
+	}
+
+	hash, err := nameHash(name)
```

Comment:
should the format of `name` be validated?

Reply:
The contract will verify the name.